File: app/assets/javascripts/pokemon.js
function Pokemon( attributes ) {
  this.name = attributes.name;
  this.maxHealth = attributes.health;
  this.health = this.maxHealth;
  this.attackStat = attributes.attackStat;
}

Pokemon.prototype = {
  pokemonList: [{
    name: "Bulbasaur",
    health: 40,
    attackStat: 15
  },
  {
    name: "Charmander",
    health: 50,
    attackStat: 12
  },
  {
    name: "Pikachu",
    health: 45,
    attackStat: 15
  },
  {
    name: "Richurtle",
    health: 80,
    attackStat: 5
  }],
  attack: function() {
    var hitRoll = Math.random();

    // If the roll is too low, the attack is a miss
    if ( hitRoll < 0.05 ) {
      return 0;
    // Most rolls will do between 75% - 100% damage.
    } else if ( hitRoll < 0.95 ) {
      return Math.floor( this.attackStat * (0.75 + Math.random() * 0.25) );
    // If the roll is high, the attack does double damage.
    } else {
      return this.attackStat * 2;
    }
  },
  receiveDamage: function( damage ) {
    this.health -= damage > this.health ? this.health : damage;
  }
};

File: lib/pokemon.rb
class Pokemon
  attr_reader :name, :max_health, :health, :attack_stat

  def initialize(attributes = {})
    @name = attributes[:name]
    @max_health = attributes[:max_health]
    @health = @max_health
    @attack_stat = attributes[:attack_stat]
  end

  def attack
    hit_roll = rand(20)
    
    # 5% chance to miss attack
    # 5% chance to do double damage
    # Most of the time, deals between 75 and 100% damage
    case hit_roll
    when 0
      return 0
    when 19
      return attack_stat * 2
    else
      damage = attack_stat * (0.75 + rand() * 0.25)
      return damage.to_i
    end
  end

  def receive_damage(damage)
    @health -= damage > @health ? @health : damage
  end
end